4단계 - 자동차 경주(우승자)

jaeykweon opened the pull request:
4단계 - 자동차 경주(우승자) 리뷰 요청드립니다.

docs는 [ReadMe](https://github.com/jaeykweon/kotlin-racingcar/tree/step4/src/main/kotlin/car_race)에서 확인하실 수 있습니다.

지난 step3에서 마지막 코드 리뷰 주신 부분도 반영하였습니다

질문 사항은 코멘트 달아놓겠습니다 감사합니다

jaeykweon committed:
refactor: [3단계] 코드리뷰 적용

- map -> forEach 변경
- 불필요한 변수 제거
- println을 view로 넘김

jaeykweon committed:
docs: [4단계] ReadMe 내용 작성

jaeykweon committed:
feature: [4단계] CarName, CarNames 구현

jaeykweon committed:
test: [4단계] CarName, CarNames 단위 테스트

jaeykweon committed:
refactor: [4단계] 차 도메인, UI, Main 함수 리팩토링

jaeykweon committed:
test: [4단계] Cars 단위 테스트 추가

jaeykweon committed:
test: [4단계] Car 테스트 리팩토링

jaeykweon commented on src/main/kotlin/car_race/logic/car/Car.kt:
```diff
@@ -3,10 +3,15 @@ package car_race.logic.car
 ) {
-    fun getPosition(): CarPosition {
-        return carPosition
+    fun getPosition() = carPosition
+
+    fun getName() = carName.getName()
```
프로퍼티에 관한 내용을 반환할 때, `getPostion`처럼 객체 자체를 반환해도 될까요 아니면 `getName`처럼 값만 반환해야할까요?
객체 자체를 반환하는 것은 getter나 다름없어보이고, 그러면 `carName`을 private으로 설정한 이유가 없어 보입니다. 어떻게 해야할까요?

ohgillwhan replied:
`CarName`과 `CarPosition`은 불변하니, getter를 없애고 외부에 그대로 노출해도 괜찮다 생각합니다. 하지만 `carPosition`은 `var`이기 때문에 `private`을 푸는것은 외부에서 수정할 수 있기 때문에 고민거리일것 같습니다. 만약 이런 케이스라면 `getNameAndPosition` 만을 통해 값에 접근할 수 있도록 하고, 더 나은 방법은 `getter`관련을 제거하는것 입니다. 대신 `nextRound` 메서드의 결과를 `Pair<CarName, CarPosition>` 을 반환하는게 좋은것 같습니다.

그리고 도메인 클래스는 어디에든 존재해도 되기 때문에 `.getName()`을 호출하지 않고 `carName`을 바로 반환해도 괜찮습니다.

ohgillwhan replied:
위의 리뷰를 고민해보시면 `Car`에서는 자연스레 `getter`는 제거되고, `private`은 유지하실 수 있을것입니다.

jaeykweon replied:
일단 Pair를 쓰는거 자체는 문제가 없나 보네요
그런데 `getNameAndPosition`만으로 값에 접근한다 하면, 이름 또는 위치 둘 중 하나만 필요한 경우에도 이 메소드를 써야하는데 상관없을까요?

ohgillwhan replied:
사실 `Pair`도 저는 선호하는 편은 아니지만, 2개의 타입이 다르기에 문제 없다고 생각했습니다. 해당 메서드를 사용해도 괜찮긴 한데, 위에 말씀드린것 처럼 `getter`관련을 제거하는것은 어떨까요? `nextRound` 메서드의 결과를 `Pair<CarName, CarPosition>` 을 반환하는게 좋은것 같습니다.

jaeykweon replied:
프로퍼티의 접근 제한자를 private으로 막았던 것을 풀면서 `getNameAndPosition` 자체가 없어졌습니다
프로퍼티에서 직접 값을 가져가도록 변경하였는데 혹시 좋지 못한 코드일까요?

ohgillwhan replied:
코틀린에서는 프로퍼티에 직접 값을 조회하는건 괜찮다고 생각합니다. 대신 리뷰에 드린것 처럼 `nextRound`가 해당 값을 반환하게 하는건 어떨까요? 다른 개발자가 해당 클래스를 사용하면, 움직인 결과를 다른 메서드를 호출해서 알아야 된다는 지식이 줄어들게 될거에요.

jaeykweon commented on src/test/kotlin/car_race/logic/car/CarsTest.kt:
```diff
@@ -0,0 +1,41 @@
+/**
+ * @see Cars
+ */
+class CarsTest : FunSpec({
```
Cars의 경우 단위테스트를 어떻게 해야할지 몰라 가장 간단하게 하였습니다.
혹시 어떤 부분이 부족할까요??

ohgillwhan commented on src/main/kotlin/car_race/logic/car/CarName.kt:
```diff
@@ -0,0 +1,17 @@
+        }
+    }
+
+    fun getName() = name
```
이런 케이스는 메서드를 제거하고, `private`을 푸는게 더 좋습니다.

ohgillwhan replied:
코틀린은 변수 선언과 동시에 `getter`/`setter` 역할을 하니 중복으로 보일 수 있습니다.

jaeykweon replied:
[refactor: [4단계] 코드리뷰 적용](https://github.com/next-step/kotlin-racingcar/pull/885/commits/07aeae0992b8cd4923a96addf3b093fdde342072)에서 적용하였습니다

ohgillwhan commented on src/main/kotlin/car_race/logic/car/CarName.kt:
```diff
@@ -0,0 +1,17 @@
+package car_race.logic.car
+
+data class CarName(
```
💯

ohgillwhan commented on src/main/kotlin/car_race/logic/car/CarNames.kt:
```diff
@@ -0,0 +1,14 @@
+) {
+
+    companion object {
+        fun from(input: String): CarNames {
+            return CarNames(
+                input.split(",").map { name -> CarName(name) }
+            )
+        }
+    }
```
팩터리 메서드 대신 `secondary constructor`는 어떨까요?

https://kotlinlang.org/docs/classes.html#secondary-constructors

jaeykweon replied:
![image](https://user-images.githubusercontent.com/30148662/202454694-824156c4-2f62-4a7f-9676-635d4265114b.png)
혹시 이 방법을 말씀하시는 걸까요?
이 방법의 경우 names가 mutablelist가 되어 값의 변경을 막을 수가 없는데, 혹시 이 방법이 틀린것일까요?

ohgillwhan replied:
아래처럼 하면 어떨까요?
```kotlin
class CarNames private constructor (val names: List<CarName>) {
    construcotr(names: String) : this(names.split(",").map { CarName(it) })
}
```

jaeykweon replied:
[refactor: [4단계] 코드리뷰 적용](https://github.com/next-step/kotlin-racingcar/pull/885/commits/07aeae0992b8cd4923a96addf3b093fdde342072)에서 적용하였습니다

ohgillwhan commented on src/main/kotlin/car_race/logic/car/Cars.kt:
```diff
@@ -7,37 +7,29 @@ class Cars(
 
     fun nextRound(movingSystem: MovingSystem) {
-        cars.map { car ->
+        cars.forEach { car ->
```
리뷰반영 💯

ohgillwhan commented on src/main/kotlin/car_race/logic/car/Cars.kt:
```diff
@@ -7,37 +7,29 @@ class Cars(
-                )
-            }
-            check(inputNumber > 0) { throw IllegalArgumentException(INVALID_TYPE_MESSAGE) }
+    companion object {
+        fun from(carNames: CarNames): Cars {
+            return Cars(
+                carNames.names.map { name ->
+                    Car(name)
+                }
+            )
```
이 부분도 `Secondary Constructor`를 써보는건 어떨까요?
https://kotlinlang.org/docs/classes.html#secondary-constructors

ohgillwhan replied:
팩터리 메서드보다 생성자로 제공하는게 더 객체 생성에 밀접한것 같습니다.

jaeykweon replied:
이펙티브 자바(코틀린)에서도 `생성자 대신 정적 팩토리 메서드를 고려하라`라고 하는데, 혹시 팩토리 메서드 패턴에 결함이 있어서 실무에 쓰이지 않는걸까요??

ohgillwhan replied:
팩터리 메서드를 사용해도 괜찮은데, 해당 코드는 충분히 생성자를 만들어서 처리할 수 있을 정도로 가볍기 때문입니다. 생성자를 제공함으로써 더 풍부한 도메인 모델이 될 수 있겠죠.

jaeykweon replied:
[refactor: [4단계] 코드리뷰 적용](https://github.com/next-step/kotlin-racingcar/pull/885/commits/07aeae0992b8cd4923a96addf3b093fdde342072)에서 적용하였습니다

ohgillwhan reviewed (changes requested):
재용님 고민해볼 리뷰 남겨봤습니다!
확인 부탁드립니다!

jaeykweon committed:
refactor: [4단계] 코드리뷰 적용

- CarNames 팩토리 생성 패턴을 secondary constructor로 변경
- Cars 팩토리 생성 패턴을 secondary constructor로 변경
- 프로퍼티 접근 제한자 해제
- 불필요한 메서드 제거 및 프로퍼티 직접 호출

ohgillwhan commented on src/main/kotlin/car_race/logic/car/Car.kt:
```diff
@@ -3,11 +3,13 @@ package car_race.logic.car
 ) {
-    fun getPosition(): CarPosition {
-        return carPosition
-    }
+    val carName: CarName = carName
+
+    var carPosition: CarPosition = carPosition
+        private set
 
     fun nextRound(movingSystem: MovingSystem) {
```
`nextRound`의 결과를 `carName`, `carPosition`을 응답을 주면 충분히 생성자에 `val`로 선언이 가능할거고, `getter`도 제거가 될것 같습니다 어떻게 생각하세요?

jaeykweon replied:
그렇게 반환하도록 바꾸는 것은 문제가 없지만, 그렇게 될 경우 매 경기마다 위치 출력을 `Cars`에서 하게 되는 걸까요?
아니면 위치출력은 `Cars`와 관련이 없고 view쪽에서 알아서 하는 걸까요?
또한 출력을 한다면 초기 값과 다음 결과를 계속해서 결과로 내보내야하는데 이를 어떻게 코드로 풀어야할지 감이 오지않습니다,,

jaeykweon replied:
그리고 nextRound의 결과를 `carNmae`과 `carPosition`으로 한다 해도 최종 우승자를 가릴때 다시 `carName`을 가져와야하는데 getter가 제거될수 없지않나요..???

ohgillwhan replied:
`CarName`, `CarPosition`은 이미 한번 값을 감싼 `Value Object`이기 때문에 `Car` 보다는 `getter`를 사용하는게 더 자연스러운것 같습니다. 필요에 따라서는 `ValueObject`는 `toString()`을 구현하면 될것 같구요.

ohgillwhan replied:
```kotlin
/** Car **/
    fun nextRound(movingSystem: MovingSystem): Pair<CarName, CarPosition> {
        if (movingSystem.canMove()) {
            carPosition = carPosition.move(1)
        }
        
        return Pair(this.carName, this.carPosition)
    }
```

```kotlin
/** Cars **/
    fun nextRound(movingSystem: MovingSystem): List<Pair<CarName, CarPosition>> {
        return cars.map { car ->
            car.nextRound(movingSystem)
        }
    }
```

```kotlin
/** Main **/
    ResultView.resultViewStart()
    for (i in 1..gameRound.value) {
        val moveResult = cars.nextRound(RandomMovingSystem)
        ResultView.printEachRound(moveResult)
    }
    ResultView.printWinners(cars) // 해당 부분도 수정
```

ohgillwhan replied:
위치출력은 `Cars`와 관련없고, `View`에서 하면 될것 같습니다

ohgillwhan commented on src/main/kotlin/car_race/logic/car/CarNames.kt:
```diff
@@ -0,0 +1,5 @@
+package car_race.logic.car
+
+class CarNames private constructor (val names: List<CarName>) {
+    constructor(input: String) : this(input.split(",").map { CarName(it) })
```
```kotlin
 constructor(input: String) : this(input.split(",").map(::CarName))
```

ohgillwhan commented on src/main/kotlin/car_race/logic/car/Cars.kt:
```diff
@@ -3,41 +3,22 @@ package car_race.logic.car
-            check(inputNumber > 0) { throw IllegalArgumentException(INVALID_TYPE_MESSAGE) }
-        }
+    fun getWinners(): List<Car> {
+        val highestScore = cars.maxOf { car -> car.carPosition.position }
+        return cars.filter { car -> car.carPosition.position == highestScore }
```
`Winner`라는 클래스가 나와서 처리해줘도 좋을것 같아요.

ohgillwhan commented on src/main/kotlin/car_race/view/InputView.kt:
```diff
@@ -1,6 +1,7 @@
 package car_race.view
```
패키지명 `carrace`는 어떤가요?

ohgillwhan reviewed (approved):
재용님 코드 잘 봤습니다!
`step5`에서 리뷰 확인 후 반영 부탁드려요~
